### PR TITLE
 Automate Delegate Reports G Suite group.

### PR DESCRIPTION
**Pull request by AlbertoPdRF**
I cherry-picked a commit from #3492 because it had exactly what I needed for this!

### CODE REVIEW

**Comment by AlbertoPdRF** [comment]
Ups, you are totally right! I'll think about how to get around that.

It's possible for people to request access to the group through the groups UI, but someone has to notice that they did so and approve them manually. Also, when someone leaves, someone has to notice that and go to remove them manually. I think that's the two problems we have to solve.

**Comment by jfly** [comment]
> but someone has to notice that they did so and approve them manually

This is only an issue for people not using gsuite accounts, right? If this is a problem we want to solve, I imagine we could add some owners to the group, and they would get notified whenever a request like this comes in.

> Also, when someone leaves, someone has to notice that and go to remove them manually

I see. It's worth noting that that's going to be a problem with any non-automated mailing lists, so we should try to have some awareness/responsibility around those.

**Comment by AlbertoPdRF** [comment]
> This is only an issue for people not using gsuite accounts, right? If this is a problem we want to solve, I imagine we could add some owners to the group, and they would get notified whenever a request like this comes in.

Yes, that could definitely be solved like this.

> I see. It's worth noting that that's going to be a problem with any non-automated mailing lists, so we should try to have some awareness/responsibility around those.

The only other non-automated list would be the Regional Organizations, right? Where we don't expect many changes whatsoever.

**Comment by AlbertoPdRF** [comment]
I think the problem of people not being Staff anymore is handled now. The only thing I don't quite like is that, in the future, we'll have on the database many people with `receive_delegate_reports: true` where this field will mean nothing.

**Comment by viroulep** [comment]
> I think the problem of people not being Staff anymore is handled now. 

I may be missing something, but I don't see in the diff where the `receive_delegate_reports` field is cleared when user is removed from a team/demoted (basically if when saving a user and `receive_delegate_reports && !any_kind_of_staff?` we should be very worried).
Maybe simply add a validation on `User` for this?

Also does this mean that any member of any team can receive delegate reports?
Shouldn't we limit this to people who [`can_view_delegate_matters?`](https://github.com/thewca/worldcubeassociation.org/blob/0b7eb4000ec11ba5040b92aa599d1ecfba8a765d/WcaOnRails/app/models/user.rb#L536)?

**Comment by viroulep** [comment]
>The problem here is that when someone is demoted from a Committee/Team the user is not edited, so adding such a validation would probably break things, unless I'm missing something.

Then implement some validation to run validation on removed/updated users?
We just very much need to have something to not end up with a user having access to reports when they shouldn't.

>Right now we allow anyone who is staff to join voluntarily the Delegate Reports group, so my proposal just goes along the same lines. This can be changed if deemed necessary, but it should be discussed among more people before deciding :P

If it is implemented like this on the website, it's very likely that we considered until now that Delegate reports were restricted to people who can view Delegate matters; if now any staff can see them, shouldn't this have been discussed among more people?

**Comment by AlbertoPdRF** [comment]
> Then implement some validation to run validation on removed/updated users?
We just very much need to have something to not end up with a user having access to reports when they shouldn't.

How would you suggest to do that? I just don't see how. Also, a user with this box thicked but who's not Staff anymore won't be able to see the reports. This box just controls who is a member of the reports group (i.e. who the reports are sent to by email), and it excludes from it people who have the checkbox thicked but are no longer staff. Who can see the reports in the website is not controlled by this.

> If it is implemented like this on the website, it's very likely that we considered until now that Delegate reports were restricted to people who can view Delegate matters; if now any staff can see them, shouldn't this have been discussed among more people?

Yes, on the website it's like this and I'm not changing that, but in the past every single Staff member was a member of the Delegates group, where all the reports were sent, so everyone had access to them, and this is the approach I'm following now. To change that, I think we should discuss if we want to follow the website's approach or what we have been doing on the Google groups.

**Comment by viroulep** [comment]
>How would you suggest to do that? I just don't see how. Also, a user with this box thicked but who's not Staff anymore won't be able to see the reports. This box just controls who is a member of the reports group (i.e. who the reports are sent to by email), and it excludes from it people who have the checkbox thicked but are no longer staff. Who can see the reports in the website is not controlled by this.

Oh I see, then I think it's *very* clunky to be able to have `receive_delegate_reports` when you are actually not allowed to receive them :p

I don't have a great suggestion right now for how to check that.
My first guess would be to add the validation I talked about on `User`, and update the removed/updated users in teams after its update [in this method](https://github.com/thewca/worldcubeassociation.org/blob/master/WcaOnRails/app/controllers/teams_controller.rb#L16) (that could be a call to some `user.clear_receive_delegate_reports_if_not_staff`, maybe with a better name, the point being to "revalidate" any user that was involved in the team update).

The user validation will make sure we can't edit a user to receive reports if they actually can't, and the "post update" in the teams controller will make sure we clear this field if they are not staff anymore.

**Comment by jfly** [comment]
@viroulep, the problem is that we actually schedule team members to have a particular end date, and currently there is no code that runs when their end date is hit.

I think if we want to keep this new column clean and accurate (and I think that is worth doing), then it's probably best to clean it up in a periodic job. I'd be OK with doing that in this same periodic job that Alberto has already added some code to. Something like calling `user.clear_receive_delegate_reports_if_not_staff` for each user with this field set *before* syncing the reports@ mailing list.

**Comment by AlbertoPdRF** [comment]
Thanks @viroulep for the help! :)

File: WcaOnRails/app/models/user.rb [hunk 1]
```diff
@@ -685,9 +719,27 @@ def editable_fields_of_user(user)
...
+
+  # This method is only called in sync_mailing_lists_job.rb, right after clear_receive_delegate_reports_if_not_staff.
+  # If used without calling clear_receive_delegate_reports_if_not_staff it might return non-current Staff members.
+  def self.delegate_reports_receivers
+    candidate_delegates = User.candidate_delegates
+    senior_delegates = User.senior_delegates
+    wqac_members = Team.wqac.current_members.includes(:user).map(&:user)
+    wrc_members = Team.wrc.current_members.includes(:user).map(&:user)
+    other_staff = User.where(receive_delegate_reports: true)
+    (candidate_delegates + senior_delegates + wqac_members + wrc_members + other_staff).uniq
```
Comment (jfly) [inline WcaOnRails/app/models/user.rb]: How do you feel about changing this to recursively use our other already defined groups? I feel like that would reduce the number of api calls we have to make to google, and would also keeps things cleaner if someone goes looking at group membership in the future. I'm thinking something like this:

```ruby
%w(
  candidates@worldcubeassociation.org
  seniors@worldcubeassociation.org
  quality@worldcubeassociation.org
  regulations@worldcubeassociation.org
) + other_staff
```

File: WcaOnRails/db/migrate/20181208145408_add_receive_delegate_reports_column_to_users_table.rb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+class AddReceiveDelegateReportsColumnToUsersTable < ActiveRecord::Migration[5.2]
+  def change
+    add_column :users, :receive_delegate_reports, :boolean, default: false
```
[the remark this answers is not in the document]
Comment (AlbertoPdRF) [reply WcaOnRails/db/migrate/20181208145408_add_receive_delegate_reports_column_to_users_table.rb]: Sure!

File: WcaOnRails/app/models/user.rb [hunk 3]
```diff
@@ -396,36 +396,58 @@ def board_member?
...
+  def results_team?
+    team_member?(Team.wrt)
+  end
+
+  def software_team?
+    team_member?(Team.wst)
+  end
+
+  # See https://www.worldcubeassociation.org/documents/motions/02.2018.1%20-%20Definitions.pdf
+  def internal_staff?
```
[the remark this answers is not in the document]
Comment (AlbertoPdRF) [reply WcaOnRails/app/models/user.rb]: Yes, they are changing, although I think I would merge them like this and change them once/if the amendments to the Motions pass.

File: WcaOnRails/app/models/user.rb [hunk 4]
```diff
@@ -685,9 +719,27 @@ def editable_fields_of_user(user)
...
+    end
     fields
   end
 
+  def self.clear_receive_delegate_reports_if_not_staff
+    User.where(receive_delegate_reports: true).reject(&:any_kind_of_staff?).map { |u| u.update(receive_delegate_reports: false) }
+  end
+
+  # This method is only called in sync_mailing_lists_job.rb, right after clear_receive_delegate_reports_if_not_staff.
+  # If used without calling clear_receive_delegate_reports_if_not_staff it might return non-current Staff members.
```
[the remark this answers is not in the document]
Comment (AlbertoPdRF) [reply WcaOnRails/app/models/user.rb]: Sure!

File: WcaOnRails/app/models/user.rb (hunk 1 above)
Comment (AlbertoPdRF) [reply WcaOnRails/app/models/user.rb]: That sounds good! I think I'm not keeping `candidates@` though, as the only purpose of that group was to be part of `reports@`.

**Comment by AlbertoPdRF** [comment]
I think I've addressed your review @jfly! :)

**Comment by jfly** [comment]
LGTM, thanks @AlbertoPdRF!

**Pull request merged**